### DISCUSSION TITLE
Topics for sktime dev days 2021

### DISCUSSION THREAD

**Discussion by mloning** [alan-turing-institute/sktime#919] [Ideas]
What are interesting topics to work on during the [sktime dev days](https://github.com/sktime/community-council/blob/main/events/sktime-dev-days.md) (https://github.com/alan-turing-institute/sktime/issues/863)? 

A few ideas:
* Documentation & tutorials
* Refactoring forecasting interface (see e.g. [base classes](https://github.com/alan-turing-institute/sktime/pull/912), [exogenous variables](https://github.com/alan-turing-institute/sktime/pull/787))
* Multivariate forecasting (predicting multiple series simultaneously) (https://github.com/sktime/enhancement-proposals/pull/4)
* Time series annotation (see e.g. https://github.com/alan-turing-institute/sktime/pull/838)

Anything else you would like to discuss, get feedback on or implement during the sktime dev days?

**Comment by patrickzib** [comment]
Hi @mloning,

what about the following topics:
- Early Time Series Classification?
- Anomaly Detection / Outlier Detection?
- Clustering
- Numba / Cython performance improvements?

Patrick

**Reply by ermshaua** [reply]
In addition, I'd like:

+ Time Series Segmentation

Arik

**Reply by fkiraly** [reply]
@patrickzib, what does "early" mean in "early time series classification"?

**Reply by patrickzib** [reply]
@fkiraly „early“ with regards to the shortest prefix of the time series needed to make accurate predictions.

**Reply by fkiraly** [reply]
@ermshaua, segmentation & outlier detection are both annotation tasks, see #260, #838, and and https://github.com/sktime/enhancement-proposals/blob/main/steps/06_time_series_annotation/step.md

Good idea, but major piece of work!
We'll need a critical mass of contributors for this.

**Comment by TonyBagnall** [comment]
Early time series classification is an area of contention at the moment :) 
https://arxiv.org/abs/2102.11487
vs
https://arxiv.org/abs/2104.13257

**Reply by fkiraly** [reply]
ah, interesting! Thanks for explaining.

Well, I definitely think it's meaningful (I guess the authors of the first paper will now forever hold a grudge against me), and it is its own task or scientific type.

You basically want classification-by-cut-off, so it's somewhere in-between series-to-series prediction and time series classification, kind of a hybrid task?

Very interesting.
Might make sense to think about this together with a refactor effort on TSC/TSR?

**Reply by fkiraly** [reply]
I notice, there's also a testing aspect in this - if the question is what the earliest timepoint where information is available, or added-information-by-timepoint

**Reply by patrickzib** [reply]
@fkiraly 
I think there are two aspects to early classification:
1) Training on the „full“ length time series
2) Testing on any prefix of the full length allowing „unsafe“ prections. 

The latter might be mapped by either introducing a new method isSafe returning True/Fale or by returning a fixed class label equal to an unsafe prediction from existing methods.

**Reply by TonyBagnall** [reply]
under this model, early classification could be a capability of a standard time series classifier, and bespoke ones such as teaser could live in the same package as the other classifiers. I think matthew has implemented teaser in tsml, so we could look to convert it, unless of course patrick wants to include it :)

**Reply by fkiraly** [reply]
yes, but how would you find the answer to `isSafe` other than by rigorous quantitative performance comparison?

**Reply by fkiraly** [reply]
any case, do the two want to open a STEP on early TSC?

**Reply by fkiraly** [reply]
my main question would be, is it a new estimator scitype, or is it the same as TSC, just applied in a certain way?

**Reply by patrickzib** [reply]
@TonyBagnall @fkiraly 
it is not quite clear to me how to handle a "not safe" decision given the current framework. Given the standard interface in sktime, one might return an invalid class label like "NaN" or a predefined class label equal to a "not safe"-decision like "-1". However, given this, it is not possible to derive the predicted unsafe label, in case one wants to trace the changes of class labels over time.

Thus, ideally the early classification interface would extend the standard classifier interface by the bespoke "isSafe"-method?

**Reply by patrickzib** [reply]
`rigorous quantitative performance`
Yes, one has to do multiple tests given different prefix lengths and e.g. test for confidences of the predicted class labels. But typically early TSC would be applied in a streaming scenario, when not all data is available from the start.

**Reply by TonyBagnall** [reply]
I would just return nan if no prediction is made. @MatthewMiddlehurst has I think done some of this in tsml?

**Reply by fkiraly** [reply]
I think we carefully need to write down the task specification, and not introduce "secret conventions"

**Reply by TonyBagnall** [reply]
I think not making a prediction generalises to a valid classifier action under any circumstances, early or not early, and nan is fairly standard indicator, or maybe throw an exception

**Reply by fkiraly** [reply]
>I think not making a prediction generalises to a valid classifier action under any circumstances, early or not early, and nan is fairly standard indicator, or maybe throw an exception

let's not get into design discussions here, that's off-topic. But we can open an issue specifically for early TSC?

**Reply by TonyBagnall** [reply]
Please do, it's not a topic I'm that interested in contributing to though

**Reply by fkiraly** [reply]
I'm only interested if someone else is - how about you, @patrickzib? You brought this up

**Reply by patrickzib** [reply]
@fkiraly @TonyBagnall 
Yes, I am about to supervise a thesis on TEASER in sktime anyways. So, it would be great to open an issue. But there are also other eTSC approaches. Would be great to see these is sktime, too!

**Reply by TonyBagnall** [reply]
thats great news Patrick, look forward to it.

**Comment by TonyBagnall** [comment]
so things I would like to see, discuss, in note format as I think of them, I'll add as I go.

1) Capabilities stored with the data
2) Refactor the data checks to separate correctness of input and the reformatting of input
3) Use scikit learn tags instead of cloned tags
4) Replace classifier capabilities with tags
5) Refactor series_as_features package, remove cloned classifiers
6) Refactor and tidy up utils directory
7) Redesign the unit tests for classifiers
8) Fix the write to ts format, generally improve the data input and output
9) Unify the benchmarking directory with out experiments.py
10) Include ability to analyse results in a similar way to tsml
11) Redesign distance functions to use 1-D arrays and numba (remove cython), including timing test
12) Complete equivalence testing vs tsml
13) Profile classifiers to find timing bottlenecks

**Reply by FlukeAndFeather** [reply]
Hi sktime team. Thanks for all your hard work! I saw on Twitter there's a "Call for sktime community presentations". Does that include projects applying sktime to domain-specific problems? If so, my collaborators and I have a project we'd like to share with the sktime community. We're applying time series classification to animal behavior analysis. Specifically, we train a hierarchical model ([stickleback](http://github.com/flukeandfeather/stickleback), work in progress) to identify behavioral events (like feeding, breathing, etc) in animal-borne sensor data. I would be happy to present our work at a dev day, if that's what you're looking for. Thanks!

**Reply by TonyBagnall** [reply]
hi, that sounds really interesting, I'm not sure what @fkiraly and @mloning had in mind. You could also consider submitting a paper to our workshop https://project.inria.fr/aaltd21/

**Reply by mloning** [reply]
Hi @FlukeAndFeather, yes, great, happy to have you present. How much time do you think you will need? I have added you to the table [here](https://github.com/sktime/community-council/blob/main/events/sktime-dev-days.md#call-for-sktime-community-presentations), hope that's okay, please feel free to create a PR to add more details!

**Reply by mloning** [reply]
Hi @FlukeAndFeather could you confirm whether you will present tomorrow?

**Comment by fkiraly** [comment]
how about: finally fixing the broken binder

**Comment by fkiraly** [comment]
python 3.9 support! #664, ping @muslimbeibytuly, @discdiver, @xuyxu 

Do we have a good view what needs to be done to make `sktime` 3.9 compatible? Can we break this down in smaller work items?

**Reply by mloning** [reply]
mainly depends on our dependencies I think and if they support it already

**Reply by fkiraly** [reply]
if that's the only thing, we could just up it and give warnings for estimator dependencies which are unsupported?

**Reply by mloning** [reply]
the CI/CD workflows don't work like that, appveyor also didn't support it yet when I last checked (see #664)

**Comment by fkiraly** [comment]
This is how the consolidated list currently looks like:
* documentation & tutorials
* forecasting - refactor, multivariate
* sktime-dl - deep learning interfaces, for TSC and forecasting
* TSC refactor & TSR
* time series clustering (new module)
* TS transformers, sktime-neuro
* annotation (new module)

**Comment by FlukeAndFeather** [comment]
Hi, sorry for not confirming sooner. We are likely going to be in the field tomorrow collecting data, so unfortunately I won’t be able to join. I’ll keep my eyes open for future dev days. Thanks!

--
Max Czapanskiy
PhD Candidate in Biology
Stanford University Hopkins Marine Station
Goldbogen Lab
________________________________

**Reply by fkiraly** [reply]
@FlukeAndFeather, here's an instance of "future dev days":
https://github.com/alan-turing-institute/sktime/discussions/2827
